File: GamePanel.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.mazegame;

/**
 *
 * @author lsuti
 */
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.*;
import javax.imageio.ImageIO;

public class GamePanel extends JPanel implements ActionListener, KeyListener {

    private javax.swing.Timer timer;
    private Player player;
    private Maze maze;
    private java.util.List<Enemy> enemies;
    private java.util.List<Collectible> collectibles;
    private java.util.List<PowerUp> powerUps;
    private Sprite exit;
    private JLabel statusLabel;

    private int timeRemaining = 60;
    private boolean collisionEnabled = false;
    private int clockTicks = 0;

    public GamePanel() {
        setPreferredSize(new Dimension(800, 600));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);

        initGame();  // create player, maze, sprites, etc.

        // now player is non-null, so this call is safe:
        statusLabel = new JLabel(getStatusText());

        timer = new javax.swing.Timer(100, this);
        timer.start();
    }

    private void initGame() {
        maze = new Maze(this, player);          // walls + exit + enemies
        player = new Player(this, maze);
        enemies = maze.getEnemies();
        exit = maze.getExit();

        collectibles = new ArrayList<>();
        powerUps = new ArrayList<>();
        setupCollectibles();
        setupPowerUps();
    }

    public JLabel getStatusLabel() {
        return statusLabel;
    }

    private void setupCollectibles() {
        Random rnd = new Random();
        int[][] layout = Maze.getMazeLayout();
        for (int i = 0; i < 3; i++) {
            int col, row;
            do {
                col = rnd.nextInt(layout[0].length);
                row = rnd.nextInt(layout.length);
            } while (layout[row][col] != 0);
            Collectible c = new Collectible(loadImage("/images/collectible.png"),
                    col * 40, row * 40);
            collectibles.add(c);
        }
    }

    private void setupPowerUps() {
        Random rnd = new Random();
        int[][] layout = Maze.getMazeLayout();
        for (int i = 0; i < 5; i++) {
            int col, row;
            do {
                col = rnd.nextInt(layout[0].length);
                row = rnd.nextInt(layout.length);
            } while (layout[row][col] != 0);
            PowerUp p;
            switch (i) {
                case 0:
                    p = new TimeReductionPowerUp(loadImage("/images/reduction.png"), col * 40, row * 40);
                    break;
                case 1:
                    p = new TimeIncreasePowerUp(loadImage("/images/increase.png"), col * 40, row * 40);
                    break;
                case 2:
                    p = new ScoreBoostPowerUp(loadImage("/images/boost.png"), col * 40, row * 40);
                    break;
                case 3:
                    p = new ShieldPowerUp(loadImage("/images/shield.png"), col * 40, row * 40);
                    break;
                default:
                    if (rnd.nextDouble() < 0.05) {
                        p = new ImmunityPowerUp(loadImage("/images/reduction.png"), col * 40, row * 40);
                    } else {
                        p = new SpeedBoostPowerUp(loadImage("/images/speed.png"), col * 40, row * 40);
                    }
            }
            powerUps.add(p);
        }
    }

    public BufferedImage loadImage(String path) {
        try {
            return ImageIO.read(getClass().getResource(path));
        } catch (IOException | NullPointerException e) {
            System.err.println("Image not found: " + path);
            return null;
        }
    }

    private String getStatusText() {
        return String.format("Time: %d  | Score: %d  | Combo: x%d  | Power-Ups Left: %d",
                timeRemaining, player.getScore(), player.getComboMultiplier(),
                (int) powerUps.stream().filter(PowerUp::isActive).count());
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2 = (Graphics2D) g;
        // draw walls
        for (Wall[] row : maze.getWalls()) {
            for (Wall w : row) {
                if (w != null) {
                    w.draw(g2);
                }
            }
        }
        // exit
        exit.draw(g2);
        // collectibles
        for (Collectible c : collectibles) {
            if (c.isActive()) {
                c.draw(g2);
            }
        }
        // powerups
        for (PowerUp p : powerUps) {
            if (p.isActive()) {
                p.draw(g2);
            }
        }
        // enemies
        for (Enemy e : enemies) {
            e.draw(g2);
        }
        // player
        player.draw(g2);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        clockTicks++;
        if (!collisionEnabled && timeRemaining < 57) {
            collisionEnabled = true;
        }

        // tick every 10 → 1s
        if (clockTicks % 10 == 0) {
            if (timeRemaining > 0) {
                timeRemaining--;
                for (Enemy en : enemies) {
                    en.update(this, collisionEnabled);
                }
                checkCollisions();
            } else {
                endGame("Time's up.");
            }
        }
        player.update(this, collisionEnabled);
        statusLabel.setText(getStatusText());
        repaint();
    }

    private void checkCollisions() {
        // power-ups
        for (PowerUp p : new ArrayList<>(powerUps)) {
            if (p.isActive() && p.getBounds().intersects(player.getBounds())) {
                p.applyEffect(player, this);
                p.setActive(false);
                player.getStatsTracker().incrementPowerUpsUsed();
            }
        }
        // collectibles
        for (Collectible c : collectibles) {
            if (c.isActive() && c.getBounds().intersects(player.getBounds())) {
                player.getInventory().addItem(c.getItemName());
                c.setActive(false);
            }
        }
        // enemies
        for (Enemy en : new ArrayList<>(enemies)) {
            if (en.getBounds().intersects(player.getBounds())) {
                if (player.isShielded()) {
                    en.setActive(false);
                    player.getStatsTracker().incrementEnemiesDefeated();
                } else if (!player.isImmune()) {
                    player.getStatsTracker().recordDeathCause("Enemy Collision");
                    endGame("Hit by an enemy.");
                    return;
                }
            } else {
                // near-death?
                if (player.getBounds().intersection(en.getBounds()).isEmpty()
                        && Math.abs(player.getX() - en.getX()) < 20
                        && Math.abs(player.getY() - en.getY()) < 20) {
                    player.getStatsTracker().recordNearDeath();
                }
            }
        }
        // exit
        if (exit.getBounds().intersects(player.getBounds())) {
            if (player.getInventory().hasItem("Key 1")) {
                endGame("You Win!");
            } else {
                System.out.println("Need Key 1 to exit!");
            }
        }
    }

    private void endGame(String msg) {
        timer.stop();
        int timeBonus = timeRemaining * getTimeScoreRatio();
        player.addScore(timeBonus);
        if (msg.equals("You Win!")) {
            player.getStatsTracker().updateFastestCompletion(60 - timeRemaining);
        }

        JOptionPane.showMessageDialog(this,
                msg + "\nFinal Score: " + player.getScore(),
                "Game Over", JOptionPane.INFORMATION_MESSAGE);

        String name = JOptionPane.showInputDialog(this,
                "Enter name for leaderboard:");
        if (name != null && !name.isBlank()) {
            Leaderboard lb = new Leaderboard();
            lb.addScore(name, player.getScore());
            lb.saveScores();
            JOptionPane.showMessageDialog(this, lb.getLeaderboard());
        }
        // achievements & stats
        player.getAchievements().checkAndUnlock(
                player.getScore(), timeRemaining,
                player.getPowerUpsCollected());
        JOptionPane.showMessageDialog(this,
                player.getAchievements().displayAchievements(),
                "Achievements", JOptionPane.INFORMATION_MESSAGE);
        JOptionPane.showMessageDialog(this,
                player.getStatsTracker().displayStats(),
                "Statistics", JOptionPane.INFORMATION_MESSAGE);
        System.exit(0);
    }

    // KeyListener
    @Override
    public void keyPressed(KeyEvent e) {
        player.keyPressed(e);
    }

    @Override
    public void keyReleased(KeyEvent e) {
        player.keyReleased(e);
    }

    @Override
    public void keyTyped(KeyEvent e) {
    }

    // expose for power-ups needing ratio/base
    public int getTimeScoreRatio() {
        return 2;
    }

    public int getScoreBoostBase() {
        return 10;
    }

    public int getTimeRemaining() {
        return timeRemaining;
    }

    public void setTimeRemaining(int t) {
        timeRemaining = t;
    }

    public JLabel getStatusLabelObj() {
        return statusLabel;
    }
}

File: GamePanel.ts
// File: src/GamePanel.ts
import { Maze } from './Maze';
import { Player } from './Player';
import { Enemy } from './Enemy';
import { Collectible } from './Collectible';
import { PowerUp } from './PowerUp';
import { ImmunityPowerUp } from './ImmunityPowerUp';
import { Inventory } from './Inventory';
import { Leaderboard } from './Leaderboard';
import { StatsTracker } from './StatsTracker';
import { Sprite } from './Sprite';
import {TimeReductionPowerUp} from "./TimeReductionPowerUp";
import {TimeIncreasePowerUp} from "./TimeIncreasePowerUp";
import {ScoreBoostPowerUp} from "./ScoreBoostPowerUp";
import {ShieldPowerUp} from "./ShieldPowerUp";
import {SpeedBoostPowerUp} from "./SpeedBoostPowerUp";

export class GamePanel {
    canvas: HTMLCanvasElement;
    ctx: CanvasRenderingContext2D;
    statusLabel: HTMLDivElement;
    private timerId: number;
    maze!: Maze;
    player!: Player;
    enemies: Enemy[] = [];
    collectibles: Collectible[] = [];
    powerUps: PowerUp[] = [];
    exit!: Sprite;
    timeRemaining: number = 60;
    collisionEnabled: boolean = false;
    private clockTicks: number = 0;
    scoreBoostBase = 10;
    timeScoreRatio = 2;

    constructor() {
        this.canvas = document.createElement('canvas');
        this.canvas.width = 800;
        this.canvas.height = 600;
        this.canvas.style.border = '2px solid #333';
        this.canvas.style.background = '#f0f0f0';

        this.ctx = this.canvas.getContext('2d')!;
        this.statusLabel = document.createElement('div');
        this.initGame();
        window.addEventListener('keydown', e => this.onKeyDown(e));
        window.addEventListener('keyup', e => this.onKeyUp(e));
        this.timerId = window.setInterval(() => this.gameLoop(), 100);
    }

    private initGame() {
        this.maze = new Maze(this);
        this.player = new Player(this, this.maze);
        this.maze.setupEnemies(this, this.player);
        this.enemies = this.maze.getEnemies();
        this.exit = this.maze.getExit();
        this.collectibles = [];
        this.powerUps = [];
        this.setupCollectibles();
        this.setupPowerUps();
    }

    private setupCollectibles() {
        const layout = this.maze.getMazeLayout();
        for (let i = 0; i < 3; i++) {
            let col: number, row: number;
            do {
                col = Math.floor(Math.random() * layout[0].length);
                row = Math.floor(Math.random() * layout.length);
            } while (layout[row][col] !== 0);

            const img = this.loadImage('/images/collectible.png');
            const itemName = `Key ${i + 1}`;        // give each collectible a name
            this.collectibles.push(
                new Collectible(img, col * 40, row * 40, itemName)
            );
        }
    }

    private setupPowerUps() {
        const layout = this.maze.getMazeLayout();

        // list out every PowerUp subclass you want to spawn
        const types = [
            ImmunityPowerUp,
            TimeReductionPowerUp,
            TimeIncreasePowerUp,
            ScoreBoostPowerUp,
            ShieldPowerUp,
            SpeedBoostPowerUp,
        ];

        // map each class name to its image path
        const imgPaths: Record<string, string> = {
            ImmunityPowerUp: '/images/reduction.png',
            TimeReductionPowerUp: '/images/reduction.png',
            TimeIncreasePowerUp: '/images/increase.png',
            ScoreBoostPowerUp: '/images/boost.png',
            ShieldPowerUp: '/images/shield.png',
            SpeedBoostPowerUp: '/images/speed.png',
        };

        const spawnCount = 6;  // how many you want out there

        for (let i = 0; i < spawnCount; i++) {
            let col: number, row: number;
            // find an empty cell, and avoid your player start (e.g. 1,1)
            do {
                col = Math.floor(Math.random() * layout[0].length);
                row = Math.floor(Math.random() * layout.length);
            } while (
                layout[row][col] !== 0 ||
                (col === 1 && row === 1)
                );

            // pick a type to spawn
            const Type = types[i % types.length];
            const img = this.loadImage(imgPaths[Type.name]);
            this.powerUps.push(new Type(img, col * 40, row * 40));
        }
    }

    loadImage(path: string): HTMLImageElement {
        const img = new Image();
        img.src = path;
        return img;
    }

    private onKeyDown(e: KeyboardEvent) {
        this.player.handleKeyDown(e);
    }

    private onKeyUp(e: KeyboardEvent) {
        this.player.handleKeyUp(e);
    }

    private gameLoop() {
        this.clockTicks++;
        if (!this.collisionEnabled && this.timeRemaining < 57) {
            this.collisionEnabled = true;
        }
        if (this.clockTicks % 10 === 0) {
            if (this.timeRemaining > 0) {
                this.timeRemaining--;
                for (const en of this.enemies) {
                    en.update(this, this.collisionEnabled);
                }
                this.checkEnemyCollisions();
                this.checkCollisions();
            } else {
                this.endGame("Time's up.");
            }
        }
        this.player.update(this, this.collisionEnabled);
        this.checkPowerUpCollisions();
        this.checkCollectibleCollisions();
        this.statusLabel.textContent = this.getStatusText();
        this.render();
    }

    private checkPowerUpCollisions() {
        for (const pu of this.powerUps) {
            if (pu.active && this.rectsIntersect(pu.getBounds(), this.player.getBounds())) {
                pu.applyEffect(this.player, this);
                pu.active = false;
                this.player.addScore(20);
                this.player.stats.incrementPowerUpsUsed();
            }
        }
    }

    private checkCollectibleCollisions() {
        for (const c of this.collectibles) {
            if (c.active && this.rectsIntersect(c.getBounds(), this.player.getBounds())) {
                this.player.getInventory().addItem(c.itemName);
                c.active = false;
                this.player.addScore(50);
            }
        }
    }

    private checkEnemyCollisions() {
        const nearThreshold = 20;
        const pRect = this.player.getBounds();
        for (const enemy of this.enemies) {
            if (!enemy.active) continue;
            const eRect = enemy.getBounds();
            // Near-death detection
            if (Math.abs(pRect.x - eRect.x) < nearThreshold && Math.abs(pRect.y - eRect.y) < nearThreshold && !this.rectsIntersect(eRect, pRect)) {
                this.player.stats.recordNearDeath();
            }
            // Actual collision
            if (this.rectsIntersect(eRect, pRect)) {
                if (this.player.isShielded()) {
                    enemy.active = false;
                    this.player.stats.incrementEnemiesDefeated();
                } else {
                    this.endGame('Hit by an enemy.');
                }
                break;
            }
        }
    }


    private rectsIntersect(a: DOMRect, b: DOMRect): boolean {
        return (
            a.x < b.x + b.width &&
            a.x + a.width > b.x &&
            a.y < b.y + b.height &&
            a.y + a.height > b.y
        );
    }

    private checkCollisions() {
        const r1 = this.exit.getBounds();
        const r2 = this.player.getBounds();
        if (
            r1.x < r2.x + r2.width &&
            r1.x + r1.width > r2.x &&
            r1.y < r2.y + r2.height &&
            r1.y + r1.height > r2.y
        ) {
            this.endGame('You Win!');
        }
    }

    private endGame(msg: string) {
        clearInterval(this.timerId);
        if (msg === 'You Win!') {
            this.player.addScore(200);
            this.player.addScore(this.timeRemaining * 5);
            const completionTime = 60 - this.timeRemaining;
            this.player.stats.updateFastestCompletion(completionTime);
        }
        alert(`${msg}\nFinal Score: ${this.player.score}`);
        const name = prompt('Enter your name for the leaderboard:');
        if (name) {
            const lb = new Leaderboard();
            lb.addScore(name, this.player.score);
            lb.saveScores();
            alert(lb.getLeaderboard());
        }
        this.player.ach.checkAndUnlock(
            this.player.score,
            this.timeRemaining,
            this.player.powerUpsCollected
        );
        alert(this.player.ach.displayAchievements());
        alert(this.player.stats.displayStats());
    }



    private getStatusText(): string {
        const left = this.powerUps.filter(p => p.active).length;
        return `Time: ${this.timeRemaining} | Score: ${this.player.score} | Combo: x${this.player.combo} | Power-Ups Left: ${left}`;
    }

    private render() {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        for (const row of this.maze.getWalls()) {
            for (const w of row) {
                if (w) w.draw(this.ctx);
            }
        }
        this.exit.draw(this.ctx);
        this.collectibles.forEach(c => c.active && c.draw(this.ctx));
        this.powerUps.forEach(p => p.active && p.draw(this.ctx));
        this.enemies.forEach(e => e.draw(this.ctx));
        this.player.draw(this.ctx);
    }
}
